### ISSUE TITLE
Test TimeoutTest.TestTimeOutCaseWithOutElapsed is failing randomly

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#789]
I am commenting out the test for the beta 4 release, we can investigate after the release.

**Update**

This test failed on CF when I was packaging beta 4. @CharliePoole couldn't reproduce, so thought it was fixed by recent changes. It then happened again on the Travis CI server for the .NET 4.5 build, so it is likely a very intermittent failure.

**Comment by CharliePoole** [comment]
@rprouse I'm not getting an error when I remove the #if NETCF condition. My VS2008 is version 9.0.30729.1 SP and it's using the Windows Mobile 5.0 Pocket PC R2 emulator. How are you running and what error are you getting?

**Comment by CharliePoole** [comment]
@rprouse Still waiting to hear from you on this.

**Comment by rprouse** [comment]
@CharliePoole, sorry I forgot :-1: 

I ran the tests in Pocket PC 2003 SE Emulator and the error is,

```xml
<test-case id="3629" name="TestTimeOutTestCaseWithOutElapsed" fullname="NUnit.Framework.Attributes.TimeoutTests.TestTimeOutTestCaseWithOutElapsed" methodname="TestTimeOutTestCaseWithOutElapsed" classname="NUnit.Framework.Attributes.TimeoutTests" runstate="Runnable" seed="1305244492" result="Failed" start-time="2015-08-25 23:55:08Z" end-time="2015-08-25 23:55:08Z" duration="0.261618" asserts="0">
  <failure>
    <message>  Expected: &lt;Failed&gt;
  But was:  &lt;Failed:Cancelled&gt;
    </message>
    <stack-trace>at NUnit.Framework.Attributes.TimeoutTests.TestTimeOutTestCaseWithOutElapsed()
    </stack-trace>
  </failure>
</test-case>
```

**Comment by CharliePoole** [comment]
Thanks Rob, I'm in full flight on a different issue right now, but I'll
come back to it and retest using that emulator.

**Comment by rprouse** [comment]
Interesting, I got the exact same failure running on Travis for .NET 4.5 today. Maybe a timing related error? I am going to re-open and rename, but probably a very low priority if it fails so infrequently.

From Travis,

```
1) Failed : NUnit.Framework.Attributes.TimeoutTests.TestTimeOutTestCaseWithOutElapsed
		  Expected: <Failed>
		  But was:  <Failed:Cancelled>
		at NUnit.Framework.Attributes.TimeoutTests.TestTimeOutTestCaseWithOutElapsed () [0x00000] in <filename unknown>:0
		Tests Not Run
```

**Comment by rprouse** [comment]
I have now seen this 4 times on Travis today while pushing other changes. I am going to up the priority back to normal because it is happening very regularly there. Interestingly, I have only seen it on Travis, not on my Linux box.

**Comment by rprouse** [comment]
@CharliePoole, my latest theory on this is that it happens on single core machines, thus the Travis and CF failures, but not on our machines. While debugging the test, I noticed that as I was stepping through `WorkItem.RunTestOnOwnThread()`, each of the two `TestCases` was stepping through at the same time and nearly in lock step.

My theory is that on a single core machine, we default to two child workers so the two `TestCases` continue to step through in parallel, but occasionally with context switching, the one test case occasionally times out.

I thought the fix might be as easy as adding `[Parallelizable(ParallelScope.None)]` to the fixture to cause the two test cases to be executed sequentially, but stepping through the code in the debugger, they were executing at the same time. This isn't the expected behavior is it, or have I misinterpreted the attribute or it's usage? The docs say In the current release, parallel execution of methods within a class is not yet implemented, but I am pretty sure the test cases are executing in parallel.

**Comment by rprouse** [comment]
I stepped through the code at a slightly higher level and `CompositeWorkItem.RunChildren` is definately creating a new thread for each of the child test cases and I don't see any code in there that respects `WorkItem.IsParallizable`. I assume this is a bug that this code is running TestCases in parallel without a way to turn it off?

**Comment by CharliePoole** [comment]
Definitely sounds like it. It could be an experiment that I mistakenly
checked in.

**Comment by oznetmaster** [comment]
Closing my issue removed the second part, which is the seeming erroneous error message itself:

Isn't the error itself a bit strange?

Shouldn't it be

9) Failed : NUnit.Framework.Attributes.TimeoutTests.TestTimeOutTestCaseWithOutElapsed
Expected: <Failed:Cancelled>
But was: <Failed>
at NUnit.Framework.Attributes.TimeoutTests.TestTimeOutTestCaseWithOutElapsed()

**Comment by oznetmaster** [comment]
Also lost was the followup (perhaps should be a separate issue?):

How can the specific TestCase which is failing be determined?

It does not seem to be available in the results.

Should it be?

**Comment by rprouse** [comment]
The error that is being reported is because we are not reporting on the test being run, but on the overall result of the run. Also, the individual results should be available as child results of the results. I would need to step through and double check though :smile:

**Comment by rprouse** [comment]
I believe this is happening because _topLevelWorkItem is null in `SimpleWorkOrderDispatcher.Dispatch(...)`. @CharliePoole, shouldn't it be set to the CompositeWorkItem that represents the parent paramterized test suite?

**Comment by rprouse** [comment]
Actually, looking at this again, I think that the two `TestCases` being run in parallel is just a symptom of the fact that we are running the test method, not running a suite. We pass the test method into the `TestBuilder.RunTest` call, so the context is not setup the same was as if we were running an entire suite or even a fixture.

I guess this could still be an issue if the user specifies a single test method on the command line.

**Comment by CharliePoole** [comment]
No, it should be set to the topmost item in the entire test run. This is
probably a second level test being run individually.

I'm on my phone so I'll have to check later.

**Comment by CharliePoole** [comment]
@oznetmaster What did you think the message should be? You apparently included angle brackets in the text, so it is being deleted here... Put the message within markdown "fences" (```) or replace the brackets with html entities.

**Comment by CharliePoole** [comment]
@rprouse Looking at the test, I don't understand why we expect the result to be Failed:Cancelled. I would think that the suite should be failed, since it contains a failed test method. And yet, it is passing at least some of the time. This may be an artifact of how TestBuilder, which we only use for running second-level tests, runs test methods.

There's a weird construction in TimeoutFixture.cs, where one of the tests is called by other tests. I don't think that's a problem, but it's confusing so I'd eliminate it by adding a private method or sleeping directly.

All in all, this looks pretty brittle. Maybe we should start over, looking at what the heck we are trying to test here!

**Comment by CharliePoole** [comment]
@oznetmaster Never mind... it's clear in #823. The result of the overall suite should, IMO be Failure, not Cancelled. The two individual tests should Succeed and Fail, respectively. Probably, each assert should have a message to make it clear what failed.

**Comment by CharliePoole** [comment]
@rprouse The apparent parallelism could be because you are looking at a test that runs a second-level test. You may want to change some names so that it's less confusing.

**Comment by CharliePoole** [comment]
The problem is definitely in the test code rather than NUnit. See PR #829 which ignores the test temporarily.

**Comment by rprouse** [comment]
Commented out in PR #829, see the comment there. I am moving this out to post-3.0 since it doesn't really block the release.

**Comment by oznetmaster** [comment]
Just want to note that it fails 100% of the time in the CF build.

**Comment by CharliePoole** [comment]
@oznetmaster Also in .NET 4.5. See comment in PR #829.

**Comment by rprouse** [comment]
I don't think we're still seeing this. Closing.
